plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,455 @@
+          deriving Eq
+
+-- | The underlying classic getter function for depth
+--
+-- Note: Top is arbitrarily defined as having depth 0.
+-- It could also be made 'minBound' or 'Word', but this is conceptually clearer and allows us to use 'IM.IntMap's.
+_depth :: Rank -> Int
+_depth = \case
+  Top -> 0
+  LamDep d _ -> d
+
+-- | Lens-style getter of the depth of a rank
+depth :: Getting r Rank Int
+depth  = to _depth
+
+instance Ord Rank where
+  compare = compare `on` _depth
+
+-- | We are interested in maximal ranks, and thus we use Rank similar to 'Data.Semigroup.Max'.
+-- OPTIMIZE: we could get rid of the following Semigroup and Monoid instances by wrapping with 'Data.Semigroup.Max'.
+instance Semigroup Rank where
+  r1 <> r2 = max r1 r2
+
+-- | The minimum rank is Top (toplevel/ no lambda/Lambda dependency).
+instance Monoid Rank where
+  mempty = Top
+
+-- | During the first pass of the AST, a reader context holds the current in-scope lambda locations (as Ranks).
+type P1Ctx = ( P1Data  --  the lambdas-in-scope that surround an expression
+             , Int     --  the current depth
+             )
+
+
+-- | During the first pass of the AST, we build an intermediate table to hold the maximum ranks of each let encountered so far.
+-- This intermediate table will be transformed right at the end of 1st pass to 'FloatData'.
+type P1Data = M.Map
+              PLC.Unique --  the identifier introduced by a let-binding
+              ( Rank --  its calculated maximum rank, i.e. where this identifier can topmost/highest float outwards to
+              , PLC.Unique   --  a principal name for this binding (needed specifically because a let-Datatype-bind can introduce multiple identifiers)
+              )
+
+-- | Before we return from the 1st pass, we transform the acccumulated 'P1Data' to something that can be more easily consumed (by the second pass).
+-- This 'FloatData' is a mapping of depth=>lambda=>{let_unique}. We consume the float-data left-to-right, sorted on depth.
+--
+-- TODO: This could instead use an 'IM.IntMap', to guarantee that the consumption is in the ordering of depth. However there was a bug when using the 'IM.minViewWithKey'
+type FloatData = [(Int, --  the depth (starting from 0, which is Top)
+                 M.Map --  a mapping of lambda-locations at this depth => to all letidentifiers to float at that depth.
+                  Rank    --  the location where the let should float to: location is a lambda name or Top
+                  (S.Set PLC.Unique) --  the let bindings that should be floated/placed under this lambda or Top
+                 )]
+
+-- | First-pass: Traverses a Term to create a mapping of every let variable inside the term ==> to its corresponding maximum rank.
+p1Term ::  forall name tyname uni a
+        . (Ord (tyname a), PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
```

Comment:
Why is `Ord (tyname a)` required? Is it possible to extract the `TypeUnique` from every `tyname` and deal with them instead, so that this constraint is not needed?